### PR TITLE
internal/report: Include heap usage in the telemetry report

### PR DESCRIPTION
**Pull request by ashutosh-narkar**
Signed-off-by: Ashutosh Narkar <anarkar4387@gmail.com>

### CODE REVIEW

File: docs/content/privacy.md [hunk 1]
```diff
@@ -4,24 +4,24 @@ kind: operations
...
 
-OPA periodically reports its version to a publicly hosted, external service. The version reports only contain the OPA
-version number (e.g., v0.12.3) and a randomly generated UUID. This feature is only applicable to the `opa run` and
-`opa version` commands.
+OPA periodically reports its version, memory usage to a publicly hosted, external service. The reports contain the OPA
+version number (e.g., v0.12.3), heap usage in bytes and a randomly generated UUID. This feature is only applicable to
+the `opa run` and `opa version` commands.
 
 In case of the `opa run` command, this feature is **ON by-default** and can be easily disabled by specifying
 the `--skip-version-check` flag. When OPA is started in either the server or repl mode, OPA calls the external service
```
Comment (srenatus) [inline docs/content/privacy.md:16]: The flag name is misleading now. Perhaps we can add `--disable-telemetry` as a new flag, and check the old flag for compat?

File: internal/report/report.go [hunk 2]
```diff
@@ -89,6 +91,10 @@ func (r *Reporter) SendReport(ctx context.Context) (*DataResponse, error) {
 	rCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
 	defer cancel()
 
+	var m runtime.MemStats
+	runtime.ReadMemStats(&m)
+	r.body["heap_usage_bytes"] = strconv.FormatUint(m.Alloc, 10)
```
Comment (srenatus) [inline internal/report/report.go:96]: ❓ Why only `Alloc` (the same as `HeapAlloc`), and not `HeapInUse` or `TotalAlloc` or `Sys`? 💭

File: internal/report/report.go [hunk 3]
```diff
@@ -89,6 +91,10 @@ func (r *Reporter) SendReport(ctx context.Context) (*DataResponse, error) {
 	rCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
 	defer cancel()
 
+	var m runtime.MemStats
```
Comment (srenatus) [inline internal/report/report.go:94]: It feels a tad sneaky do this in "SendReport". At least we should add that this happens to the functions comment, if we don't find a better place to put this in.

**Review by srenatus — COMMENTED** [review]
I'm uncertain about selecting just this one field. Is it good enough to help, or do we need any of the others, too?

File: internal/report/report.go (hunk 2 above)
Comment (ashutosh-narkar) [reply internal/report/report.go:96]: Heap usage is a good indicator of OPA's memory consumption. If you think any other memory stats are useful, happy to add them as well.

Comment (srenatus) [reply internal/report/report.go:96]: I'm not sure! Not my strong suit. 

```
	// HeapSys is bytes of heap memory obtained from the OS.
	//
	// HeapSys measures the amount of virtual address space
	// reserved for the heap. This includes virtual address space
	// that has been reserved but not yet used, which consumes no
	// physical memory, but tends to be small, as well as virtual
	// address space for which the physical memory has been
	// returned to the OS after it became unused (see HeapReleased
	// for a measure of the latter).
	//
	// HeapSys estimates the largest size the heap has had.
	HeapSys [uint64](https://pkg.go.dev/builtin#uint64)
```

_"HeapSys estimates the largest size the heap has had."_ sounds interesting. 🤔

File: docs/content/privacy.md (hunk 1 above)
Comment (ashutosh-narkar) [reply docs/content/privacy.md]: Added the new `--disable-telemetry` flag.

File: internal/report/report.go (hunk 3 above)
Comment (ashutosh-narkar) [reply internal/report/report.go:95]: Updated the function comment. `SendReport` is periodically called by the runtime. So the report will contain the latest memory usage of the instance.

File: cmd/run.go [hunk 4]
```diff
@@ -193,7 +194,15 @@ To skip bundle verification, use the --skip-verify flag.
 	addConfigOverrides(runCommand.Flags(), &cmdParams.rt.ConfigOverrides)
 	addConfigOverrideFiles(runCommand.Flags(), &cmdParams.rt.ConfigOverrideFiles)
 	addBundleModeFlag(runCommand.Flags(), &cmdParams.rt.BundleMode, false)
+
 	runCommand.Flags().BoolVar(&cmdParams.skipVersionCheck, "skip-version-check", false, "disables anonymous version reporting (see: https://www.openpolicyagent.org/docs/latest/privacy)")
+	err := runCommand.Flags().MarkDeprecated("skip-version-check", "\"skip-version-check\" is deprecated. Use \"disableTelemetry\" instead")
```
Comment (srenatus) [inline cmd/run.go:199]:
```suggestion
	err := runCommand.Flags().MarkDeprecated("skip-version-check", "\"skip-version-check\" is deprecated. Use \"disable-telemetry\" instead")
```

File: cmd/run.go [hunk 5]
```diff
@@ -271,7 +280,12 @@ func initRuntime(ctx context.Context, params runCmdParams, args []string) (*runt
 		Ignore: params.ignore,
 	}.Apply
 
-	params.rt.EnableVersionCheck = !params.skipVersionCheck
+	params.rt.EnableVersionCheck = !params.disableTelemetry
+
+	// For backwards compatibility, check if `--skip-version-check` flag set.
+	if params.skipVersionCheck {
+		params.rt.EnableVersionCheck = !params.skipVersionCheck
```
Comment (srenatus) [inline cmd/run.go:287]:
```suggestion
		params.rt.EnableVersionCheck = false
```

**Review by srenatus — APPROVED** [review]
Thanks! Just nitpicks.

File: docs/content/privacy.md [hunk 6]
```diff
@@ -4,24 +4,24 @@ kind: operations
...
+version number (e.g., v0.12.3), heap usage in bytes and a randomly generated UUID. This feature is only applicable to
+the `opa run` and `opa version` commands.
 
 In case of the `opa run` command, this feature is **ON by-default** and can be easily disabled by specifying
-the `--skip-version-check` flag. When OPA is started in either the server or repl mode, OPA calls the external service
-on a best-effort basis and shares the version it's running. The time taken to execute the remote call and process the
-subsequent response from the external service does not delay OPA's start-up.
+the `--disable-telemetry` flag. When OPA is started in either the server or repl mode, OPA calls the external service
+on a best-effort basis and shares the version it's running and current memory usage. The time taken to execute the
+remote call and process the subsequent response from the external service does not delay OPA's start-up.
```
Comment (tsandall) [inline docs/content/privacy.md:18]: This seems fine for now, however, in the future we could imagine reporting other anonymous usage statistics and then we'll have to go back and edit all these sentences. Perhaps we should say "OPA periodically it's version and specific anonymous runtime statistics to a publicly..." and then define what those statistics are in a bulleted list below.

Comment (ashutosh-narkar) [reply docs/content/privacy.md]: Updated the doc to include a list of reported stats.

File: cmd/run.go (hunk 5 above)
Comment (ashutosh-narkar) [reply cmd/run.go]: Updated.

**Pull request merged**